File: LeetCode/Math/Golang/Leetcode478-generate-random-point-in-a-circle.go
package sbgo

import "math/rand"

type Solution struct {
	x, y, r float64
}

func Constructor(radius float64, x_center float64, y_center float64) Solution {
	return Solution{
		x: x_center,
		y: y_center,
		r: radius,
	}
}

func (this *Solution) RandPoint() []float64 {
	a, b := rand.Float64()*2-1, rand.Float64()*2-1
	if a*a+b*b > 1 {
		return this.RandPoint()
	}
	return []float64{
		this.x + this.r*a, this.y + this.r*b,
	}
}

File: LeetCode/Math/Rust/Leetcode478-generate-random-point-in-a-circle.rs
struct Solution {
    x: f64,
    y: f64,
    r: f64,
}

use rand::prelude::*;

impl Solution {
    fn new(radius: f64, x_center: f64, y_center: f64) -> Self {
        Self {
            x: x_center,
            y: y_center,
            r: radius,
        }
    }

    fn rand_point(&self) -> Vec<f64> {
        let mut rng = rand::thread_rng();
        let a = rng.gen_range(-1.0, 1.0);
        let b = rng.gen_range(-1.0, 1.0);
        if a * a + b * b > 1_f64 {
            return self.rand_point();
        }
        vec![self.x + self.r * a, self.y + self.r * b]
    }
}
